Add debugging guide (#6425)
fleetdm/fleet: docs/Deploying/Debugging.md

# Debugging

## Goals of this guide

This is NOT meant to be an exhaustive list of possible issues in Fleet and how to solve them.

This is a guide for going from a vague statement such as "things are not working correctly" to a more narrowed down and 
specific assessment. This doesn't mean necessarily a solution; but, with a more specific assessment, it'll be easier for 
the Engineering team to help.

Note that even if you do all your homework, the Engineering team might have follow-up questions.

## Basic data that is needed

While it's not needed strictly 100% of the times, in most cases it's extremely useful to have a clear understanding of 
the basic characteristics of the Fleet deployment with the issues:

- Amount of total hosts.
- Amount of online hosts.
- Amount and size (CPU/Mem) of the Fleet instances.
- Fleet instances CPU and Memory usage while the issue has been happening.
- MySQL flavor/version in use.
- MySQL database size (CPU/Mem).
- MySQL CPU and Memory usage while the issue has been happening.

> - Are database readers configured? If so, how many?

Reviewer: Is this another name for "read replicas"?

Author: Yeah, I should just use read replicas.

Depending on the API, there will likely be followup questions about amount of data, but this would be a good point to 
check in with Engineering.

### MYSQL

Most of the data needed to understand an issue in MySQL should've been gathered already by the basic data specified at 
the beginning of this document. However, there is a chance that Fleet is running with a database user that is not 
capable of querying the information needed. So here are the queries that would output a good first step in information
gathering:

```sql
show engine innodb status;
show processlist;
```

If read replicas are configured, another piece of important data is whether there has been any replication lag 
registered.

With all this gathered, it's a good time to reach out to the Engineering team.

### REDIS

In most cases, the data gathered at the beginning of this document should be enough to understand what might be 
happening with Redis. However, if more details are needed, running the 
[monitor command](https://redis.io/commands/monitor/) should shed more light in the issue.

**WARNING**: if Redis is suffering from performance issues, running monitor will only increase the problem.

### OSQUERY

> Just like with the Fleet server, the best way to understand issues on the client side is to look at logs.

Reviewer: Maybe explain where to look for such logs:
- Orbit: https://github.com/fleetdm/fleet/blob/main/orbit/README.md#logs
- Vanilla osquery: `/var/log/osquery/`

If you are running vanilla osquery in the host, please restart the host with `--tls_dump` and `--verbose`. This will 
allow us to see more details as to what's happening in the communication with Fleet (or lack there of).

If you are running Orbit, you should add `--debug` to the command line options. This will get debug logs for Orbit and 
also for osquery automatically.

If you are running Fleet Desktop there's no change needed, you should see the log file in the following directories 
depending on the platform:

- Linux: `$XDG_STATE_HOME/Fleet` or `$HOME/.local/state/Fleet`
- macOS: `$HOME/Library/Logs/Fleet`
- Windows: `%LocalAppData%/Fleet`

The log file name is `fleet-desktop.log`.

If the issue is related to osquery extensions, the following data would be needed:

- osquery version
- OS it's running on
- What does the extension do?
- How is the extension queried/deployed?
- What language is the extension implemented in?

With this data, it's time to reach out to Engineering.

### `INFRA`

> At this level, what you want to look into are Load Balancer logs, errors, and configurations. For instance, does the LB 

Reviewer 2: Would it make sense to have a link to https://health.aws.amazon.com/health/status (for AWS if you are logged in this will also redirect to a specific account health portal as well) and also a general "check your infra statuspage" type thing to cover other clouds?  Sometimes people don't even think about their cloud provider possibly having issues, and this lets them at least cross that off the list.

# Debugging

## Goals of this guide

This is NOT meant to be an exhaustive list of possible issues in Fleet and how to solve them.

This is a guide for going from a vague statement such as "things are not working correctly" to a more narrowed down and 
specific assessment. This doesn't mean necessarily a solution; but, with a more specific assessment, it'll be easier for 
the Engineering team to help.

> Note that even if you do all your homework, the Engineering team might have follow-up questions.

Reviewer 3: A little tweak just to soften the tone?

```suggestion
Note that even if you follow all those steps, the Engineering team might have follow-up questions.
```

## Basic data that is needed

While it's not needed strictly 100% of the times, in most cases it's extremely useful to have a clear understanding of 
the basic characteristics of the Fleet deployment with the issues:

- Amount of total hosts.

> - Amount of online hosts.

Reviewer: How about approximate amount of scheduled queries and policies.

Reviewer: Also, I wonder if we can make`fleetctl debug archive` answer all these questions for us... (have it generate a report.txt or something with number of users, etc.)

- Amount and size (CPU/Mem) of the Fleet instances.
- Fleet instances CPU and Memory usage while the issue has been happening.
- MySQL flavor/version in use.

> - MySQL database size (CPU/Mem).

Reviewer 3: "MySQL server capacity (CPU/Mem)"?

- MySQL CPU and Memory usage while the issue has been happening.
- Are database readers configured? If so, how many?

> - Redis version and size (CPU/Mem).

Reviewer 3: ```suggestion
- Redis version and server capacity (CPU/Mem).
```

[Here's how to accomplish this using Google Chrome](https://developer.chrome.com/docs/devtools/network/).

Besides from this, it might be good to continue with `MYSQL` and `REDIS`.

Depending on the API, there will likely be followup questions about amount of data, but this would be a good point to 
check in with Engineering.

### MYSQL

Most of the data needed to understand an issue in MySQL should've been gathered already by the basic data specified at 
the beginning of this document. However, there is a chance that Fleet is running with a database user that is not 
capable of querying the information needed. So here are the queries that would output a good first step in information
gathering:

```sql
show engine innodb status;
show processlist;
```

If read replicas are configured, another piece of important data is whether there has been any replication lag 
registered.

With all this gathered, it's a good time to reach out to the Engineering team.

### REDIS

In most cases, the data gathered at the beginning of this document should be enough to understand what might be 
happening with Redis. However, if more details are needed, running the 

> [monitor command](https://redis.io/commands/monitor/) should shed more light in the issue.

Reviewer 3: If using AWS Elasticache, looking at the metrics of the redis instance/cluster could be useful, looking for any metric that looks close to a limit (you've already mentioned memory and cpu usage at the start, but other metrics like current connections, replication lag if applicable, if one instance is largely overused compare to others in cluster mode, number of commands per key type could help identify what is wrong).
